File: Project Templates/TenPearls/iOS Starter App.xctemplate/objc/Controls/NibDesignable.m
//___FILEHEADER___

#import "NibDesignable.h"
#import <objc/runtime.h>

@implementation NibDesignable

- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder {

  self = [super initWithCoder:aDecoder];

  if (self && self.subviews.count == 0) {

    [self setupNib];
  }
  return self;

}

- (id)initWithFrame:(CGRect)frame {

  self = [super initWithFrame:frame];
  if (self && self.subviews.count == 0) {

    [self setupNib];
  }
  return self;
}

- (void)awakeFromNib {

  [super awakeFromNib];

  [self setupDynamicProperties];

}

- (void)setupDynamicProperties {

}

- (void)setupNib {

  UIView *view = [self loadNib];
  view.frame = self.bounds;
  view.autoresizingMask =
      UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;

  [self addSubview:view];

}

- (UIView *)loadNib {

  NSBundle *bundle = [NSBundle bundleForClass:[self class]];
  return [bundle loadNibNamed:[self nibName] owner:self options:nil][0];

}

- (NSString *)nibName {

  NSString *strClass = NSStringFromClass([self class]);
  return strClass;
}

@end

File: Project Templates/TenPearls/Tenpearls iOS Starter App.xctemplate/swift/Controls/NibDesignable.swift
//___FILEHEADER___

import UIKit
@IBDesignable
class NibDesignable: BaseView {

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        
        nibSetup()
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        nibSetup()
    }
}

extension NibDesignable {
    
    func nibSetup() {
        
        let view = loadViewFromNib()
        view.frame = bounds
        view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        view.translatesAutoresizingMaskIntoConstraints = true
        
        addSubview(view)
    }
    
    func loadViewFromNib() -> UIView {
        
        let bundle = Bundle(for: type(of: self))
        let nib = UINib(nibName: String(describing: type(of: self)), bundle: bundle)
        let nibObject: Any? = nib.instantiate(withOwner: self, options: nil).first
        
        guard let nibView: UIView = nibObject as? UIView, nibObject is UIView else {
            fatalError("nib could not be loaded")
        }
        
        return nibView
    }
}
